mutating webhook: audit log mutation existence and patch

roycaihw opened the pull request:
**Issue:**

Mutating webhooks can silently mutate objects and cause API requests to fail validation (#62666). Currently when a request goes through admission chain, kube-apiserver doesn’t log / store trace of 1. which mutating webhooks mutated the request; 2. what was changed in the request.

Moreover, when a mutated request fails validation, the validation error message doesn’t necessarily reveal the actual cause (mutating webhooks) for the request to be invalid (https://github.com/kubernetes/kubernetes/issues/65569#issuecomment-403615085). This makes it hard to debug a Kubernetes cluster, when (misconfigured) mutating webhooks silently invalid valid requests and put the cluster in undesired states.

**What this PR does:** 

This PR records name of mutating webhooks in audit log to help answer the question "which mutating webhooks mutated the request".

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Audit events now log the existence and patch of mutating webhooks. 
* At Metadata audit level or higher, an annotation with key "mutation.webhook.admission.k8s.io/round_{round idx}_index_{order idx}" gets logged with JSON payload indicating a webhook gets invoked for given request and whether it mutated the object or not.
* At Request audit level or higher, an annotation with key "patch.webhook.admission.k8s.io/round_{round idx}_index_{order idx}" get logged with the JSON payload logging the patch sent by a webhook for given request.
```

/sig api-machinery

fedebongio commented:
/cc @tallclair 
could you take a look to this one please?

roycaihw commented:
/cc @caesarxuchao 

I'm adding an e2e test

roycaihw commented:
/retest

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -158,6 +159,18 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta
+		//     "admission.k8s.io" is the name of the org,
+		//     webhook name (h.Name) is the key name,
+		//     "mutation" is the value
+		if err := attr.AddAnnotation("webhooktrace.admission.k8s.io/"+h.Name, "mutation"); err != nil {
```
Should we distinguish between metalevel auditing and object level? In the latter case we might even want to see the actual patch.

sttts replied:
would prefer `mutation.webhook..admission.k8s.io/name: true/false`.

roycaihw replied:
> Should we distinguish between metalevel auditing and object level? In the latter case we might even want to see the actual patch.

good point. I'm looking at if that's achievable here

> Is this blocked on anything?

need to rebase on reinvocation PR https://github.com/kubernetes/kubernetes/pull/78080 to treat the second pass specially. According to `AddAnnotation` doc, overwriting existing key with new value results in an error

roycaihw replied:
> Should we distinguish between metalevel auditing and object level? In the latter case we might even want to see the actual patch.

done. Adding/updating integration/e2e tests to verify behavior under different audit levels

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -158,6 +159,18 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta
+		//     "admission.k8s.io" is the name of the org,
+		//     webhook name (h.Name) is the key name,
+		//     "mutation" is the value
+		if err := attr.AddAnnotation("webhooktrace.admission.k8s.io/"+h.Name, "mutation"); err != nil {
```
Do we want/need the same for rejection of objects in the validation phase (and also here) ?

roycaihw replied:
by validation phase do you mean the `if !response.Response.Allowed` block? 

I think other code paths that return errors should already have error message clear enough to indicate the problem is related to webhook.

sttts commented:
Is this blocked on anything?

sttts commented:
`staging/src/k8s.io/apiserver/pkg/audit/request.go:230:28: "presist" is a misspelling of "persist"`

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -42,6 +43,15 @@ import (
 )
 
+const (
+	// patchAuditAnnotationPrefix is a prefix for presisting webhook patch in audit annotation.
```
persisting

roycaihw commented:
@sttts Added integration test. The test code is ugly but have it here to demonstrate it's working for now. PTAL

roycaihw commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -174,6 +184,26 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, invocatio
+	if !bytes.Equal(objJS, patchedJS) {
+		mutation = "true"
+		if err := attr.Attributes.AddAnnotation(patchAuditAnnotationPrefix+annotationKey, string(patchJS)); err != nil {
+			klog.Warningf("failed to set patch annotation for mutating webhook %s", annotationKey)
```
I'm trying to avoid logging the actual patch in apiserver log, to prevent potentially leaking PII from here

cc @liggitt

sttts commented:
Needs a better release note with the actual annotation keys.

sttts commented on test/integration/master/audit_test.go:
```diff
@@ -269,3 +344,134 @@ func expectNoError(t *testing.T, err error, msg string) {
 	}
 }
+
+// the following functions are copied from test/integration/apiserver/admissionwebhook
```
copying this code is a blocker for the PR. Please move code first where is can be reused.

sttts commented on test/integration/master/audit_test.go:
```diff
@@ -164,62 +173,128 @@ rules:
-		t.Fatalf("Failed to close audit policy file: %v", err)
-	}
+func testAudit(t *testing.T, version string, level auditinternal.Level, enableMutatingWebhook bool) {
+	t.Run(fmt.Sprintf("%s.%s.%t", version, level, enableMutatingWebhook), func(t *testing.T) {
```
move this `t.Run` into the loop above. That avoid the ugly indention.

sttts commented on test/integration/master/audit_test.go:
```diff
@@ -164,62 +173,128 @@ rules:
+				t.Fatal(err)
+			}
+			// wait for webhook config to take effect
+			time.Sleep(5 * time.Second)
```
this asks for flakes. You have to poll you test below until the annotation is visible.

sttts commented on test/integration/master/audit_test.go:
```diff
@@ -164,62 +173,128 @@ rules:
+		var url string
+		var closeFunc func()
+		if enableMutatingWebhook {
+			url, closeFunc = newWebhookServer(t)
```
defer closeFunc()

sttts commented on test/integration/master/audit_test.go:
```diff
@@ -164,62 +173,128 @@ rules:
+		if e.Verb != "create" && e.Verb != "update" {
+			continue
+		}
+		// TODO: the maps can be mutated by other test runs, because maps are copies of reference
```
which maps?

roycaihw replied:
`WebhookMutationAnnotations` and `WebhookPatchAnnotations` in `test/utils.AuditEvent`

sttts commented on test/integration/master/audit_test.go:
```diff
@@ -164,62 +173,128 @@ rules:
+			t.Fatalf("Unexpected error: %v", err)
+		}
+		defer stream.Close()
+		expected := prepareExpectedEvents(expectedEvents, level, enableMutatingWebhook)
```
never mutate globals in tests. This asks for trouble. Create them here locally in the test func instead.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/testcase.go:
```diff
@@ -517,10 +517,14 @@ func NewMutatingTestCases(url *url.URL) []Test {
+			ExpectAnnotations: map[string]string{
+				"removelabel.example.com/key1":                              "value1",
+				"mutation.webhook.admission.k8s.io/removelabel.example.com": "true",
+				"patch.webhook.admission.k8s.io/removelabel.example.com":    `[{"op": "remove", "path": "/metadata/labels/remove"}]`,
```
you had some reinvocation code. How will it look like for the second (or third) invocation?

roycaihw replied:
the key will be suffix with index to distinguish the round
```
mutation.webhook.admission.k8s.io/removelabel.example.com-1: true
mutation.webhook.admission.k8s.io/removelabel.example.com-2: false
patch.webhook.admission.k8s.io/removelabel.example.com-1: ...
```

sttts commented on staging/src/k8s.io/apiserver/pkg/audit/request.go:
```diff
@@ -40,6 +41,12 @@ import (
+	// Audit handler decides whether annotation with this prefix should be logged based on audit level.
+	// Since mutating webhook patches the request body, audit level must be greater or equal to Request
+	// for the annotation to be logged
+	patchAuditAnnotationPrefix = "patch.webhook.admission.k8s.io/"
```
having that constant twice is not good

sttts commented on staging/src/k8s.io/apiserver/pkg/audit/request.go:
```diff
@@ -220,6 +227,11 @@ func LogAnnotation(ae *auditinternal.Event, key, value string) {
 	}
+	// In mutating webhooks we persist webhook patch in audit annotation with patchAuditAnnotationPrefix.
+	// Since the patch mutates request body, we skip logging if audit level is less than Request.
+	if strings.HasPrefix(key, patchAuditAnnotationPrefix) && ae.Level.Less(auditinternal.LevelRequest) {
```
having knowledge about webhooks here is strange. something is wrong with the code layering.

roycaihw replied:
I'm working on associate each audit annotation value with an audit level property
```
annotations     map[string]struct{auditinternal.Level,string}
```

https://github.com/kubernetes/kubernetes/blob/9577cbfb9fd607d3698e1f2440f5875c85eca384/staging/src/k8s.io/apiserver/pkg/admission/attributes.go#L45

sttts commented on staging/src/k8s.io/apiserver/pkg/audit/request.go:
```diff
@@ -220,6 +227,11 @@ func LogAnnotation(ae *auditinternal.Event, key, value string) {
 		return
 	}
+	// In mutating webhooks we persist webhook patch in audit annotation with patchAuditAnnotationPrefix.
+	// Since the patch mutates request body, we skip logging if audit level is less than Request.
```
I cannot follow. Why do we change generic `LogAnnotation` logic because some plugin (webhook) does something?

sttts commented on test/integration/master/audit_test.go:
```diff
@@ -205,21 +251,86 @@ func testAudit(t *testing.T, version string) {
+		defer stream.Close()
+		missingReport, err := utils.CheckAuditLines(stream, getExpectedEvents(level, enableMutatingWebhook), versions[version])
+		if err != nil {
+			return false, fmt.Errorf("Unexpected error: %v", err)
```
golang'ism: lower-case errors

sttts commented on test/integration/master/audit_test.go:
```diff
@@ -205,21 +251,86 @@ func testAudit(t *testing.T, version string) {
+	if level.GreaterOrEqual(auditinternal.LevelRequest) {
+		webhookPatchAnnotations = map[string]string{}
+		webhookPatchAnnotations[utils.PatchAuditAnnotationPrefix+"auditmutation.integration.test"] = `[{"op":"add","path":"/foo","value":"test"}]`
+		requestObject = true
```
some godoc oneliners above each block would help to read this. This is a lot of test value construction.

sttts commented on test/utils/admission_webhook.go:
```diff
@@ -0,0 +1,107 @@
+	"k8s.io/api/admission/v1beta1"
+)
+
+func NewWebhookServer(handler http.Handler) (string, func(), error) {
```
godoc

sttts commented on test/utils/admission_webhook.go:
```diff
@@ -0,0 +1,107 @@
+	return webhookServer.URL, webhookServer.Close, nil
+}
+
+func WebhookHandler(t *testing.T, admit func(*v1beta1.AdmissionReview) error) http.HandlerFunc {
```
godoc

sttts commented on test/utils/audit.go:
```diff
@@ -30,6 +31,11 @@ import (
 	"k8s.io/apiserver/pkg/audit"
 )
 
+const (
```
same constant a third time

sttts commented on test/utils/audit.go:
```diff
@@ -46,6 +52,52 @@ type AuditEvent struct {
+}
+
+// auditEvent is a private wrapper on top of AuditEvent used by auditEventTracker
+type auditEvent struct {
```
golang'ism: helpers are usually at the bottom. makes reading easier.

sttts commented on test/utils/audit.go:
```diff
@@ -46,6 +52,52 @@ type AuditEvent struct {
+}
+
+// Match marks the given event as found if it's expected
+func (t *auditEventTracker) Match(event AuditEvent) {
```
this does not match, it marks. The verb should be the first word in a go func with side effect.

roycaihw commented:
rebased on top of webhook reinvocation. I'm adding godoc and waiting for test results

roycaihw commented:
updated godoc and fixed gofmt failure

feedback on the interface plumbing is welcome: https://github.com/kubernetes/kubernetes/pull/77824/commits/86b7b2513588b933093271e86a4ca0f747c5acf2#diff-5a87dbaca548da72585af69f20b8338fR70 (and in which package should we place the [pair type](https://github.com/kubernetes/kubernetes/pull/77824/commits/86b7b2513588b933093271e86a4ca0f747c5acf2#diff-d1848419e65f42cc9cc5f53e328f4856R48))

note that although the audit logging works with reinvocation, currently the integration test only exercises audit-logging the first pass. It would be better to share reinvocation testcases to exercise audit-logging the second pass

roycaihw commented:
/retest

sttts commented on hack/.golint_failures:
```diff
@@ -423,6 +423,7 @@ staging/src/k8s.io/apimachinery/pkg/watch
 staging/src/k8s.io/apiserver/pkg/admission
 staging/src/k8s.io/apiserver/pkg/admission/configuration
 staging/src/k8s.io/apiserver/pkg/admission/initializer
+staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook
```
can we fix the warnings?

roycaihw replied:
the warning is from rebasing to splitting webhook type
```
staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/accessors.go|25 col 6 warning| type name will be used as webhook.WebhookAccessor
```

jpbetz replied:
This was done in the Webhook API type split PR. The warning we silenced is that `webhook.WebhookAccessor` stutters, @liggitt and I were in favor of keeping the `WebhookAccessor` name instead of renaming it to just `Accessor` or something. Happy to revisit this.

sttts replied:
just call it Accessor. But that's something for @jpbetz as a followup, not htis PR.

sttts commented on staging/src/k8s.io/apiserver/pkg/audit/types.go:
```diff
@@ -44,3 +44,8 @@ type Backend interface {
 	String() string
 }
+
+type Annotation struct {
```
this type is at the wrong spot. It belongs into the admission code. For audit any annotation is the same.

roycaihw replied:
I agree. I first put it under `k8s.io/apiserver/pkg/admission` and let `k8s.io/apiserver/pkg/audit` imports it. It turns out to create `k8s.io/apiserver/pkg/admission` -> `k8s.io/apiserver/pkg/audit` -> `k8s.io/apiserver/pkg/admission` dependency cycle

I will try other places under admission (something like `.../admission/util`?)

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/interfaces.go:
```diff
@@ -63,7 +63,13 @@ type Attributes interface {
+	// AddAnnotationWithLevel sets annotation according to key-value pair with additional intended audit level.
+	// An Annotation gets logged into audit event if the request's audit level is greater or equal to the
+	// intended audit level.
 	AddAnnotationWithLevel(key, value string, level auditinternal.Level) error
```
do we need two funcs?

roycaihw replied:
otherwise we need to update places invoking `AddAnnotation()`, and they need to import `auditinternal` package. I'd like those places (test code, some admission controller package) to be unaware of `auditinternal`

sttts commented on staging/src/k8s.io/apiserver/pkg/audit/request.go:
```diff
@@ -216,21 +216,21 @@ func encodeObject(obj runtime.Object, gv schema.GroupVersion, serializer runtime
 
 // LogAnnotation fills in the Annotations according to the key value pair.
-func LogAnnotation(ae *auditinternal.Event, key, value string, level auditinternal.Level) {
+func LogAnnotation(ae *auditinternal.Event, key string, value Annotation) {
```
as my comment above, this Annotation concept is unnecessary and should not live in audit anyway.

sttts commented:
Get rid of the `Annotation` concept, and just pass key, value and level around. Then this looks good.

roycaihw commented:
moved `Annotation` concept to `admission` package. I'm adding a commit to unify AddAnnotation and AddAnnotationWithLevel

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -44,7 +43,7 @@ type attributesRecord struct {
 	// other elements are always accessed in single goroutine.
 	// But ValidatingAdmissionWebhook add annotations concurrently.
-	annotations     map[string]audit.Annotation
+	annotations     map[string]Annotation
```
why do we need this as a public concept? It's fine here as a private type. But I would not use it in any func signature.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/interfaces.go:
```diff
@@ -93,13 +93,13 @@ type ObjectInterfaces interface {
 // this interface.
 type AnnotationsGetter interface {
-	GetAnnotations() map[string]Annotation
+	GetAnnotations(level auditinternal.Level) map[string]Annotation
```
maxLevel is better maybe.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -111,35 +112,44 @@ func (record *attributesRecord) GetUserInfo() user.Info {
 		return err
 	}
-
+	if level != auditinternal.LevelMetadata && level != auditinternal.LevelRequest {
```
you can compare with <, no need to enumerate the level which are ok

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/interfaces.go:
```diff
@@ -85,13 +93,23 @@ type ObjectInterfaces interface {
+}
+
+// Annotation holds a pair of intended audit level and annotation string
+type Annotation struct {
```
private

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -120,26 +131,35 @@ func (record *attributesRecord) getAnnotations() map[string]string {
 	return cp
 }
 
 func (record *attributesRecord) AddAnnotation(key, value string) error {
```
this might need a godoc comment

sttts replied:
doesn't look like as it existed before. nvmd

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -120,26 +131,35 @@ func (record *attributesRecord) getAnnotations() map[string]string {
 	}
-
+	if level.Less(auditinternal.LevelMetadata) || level.GreaterOrEqual(auditinternal.LevelRequestResponse) {
+		return fmt.Errorf("admission annotations are not allowed to be set at level other than Metadata and Request, key: %q, level: %s", key, level)
```
is this really an error or should it just be ignored?

sttts replied:
`level.Less(auditinternal.LevelMetadata)` is enough. No need to compare with anything bigger than request.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/interfaces.go:
```diff
@@ -61,8 +62,15 @@ type Attributes interface {
 	// Both ValidationInterface and MutationInterface are allowed to add Annotations.
+	// By default, an annotation gets logged into audit event if the request's audit level is greater or
+	// equal to Metadata.
 	AddAnnotation(key, value string) error
```
we could deprecate that and clean up the callers in an follow-up

sttts commented on test/utils/admission_webhook.go:
```diff
@@ -0,0 +1,111 @@
+
+// WebhookHandler creates a HandlerFunc that decodes/encodes AdmissionReview and performs
+// given admit function
+func WebhookHandler(t *testing.T, admit func(*v1beta1.AdmissionReview) error) http.HandlerFunc {
```
AdmissionWebhookHandler

sttts commented on test/utils/admission_webhook.go:
```diff
@@ -0,0 +1,111 @@
+
+// NewWebhookServer sets up a webhook server with TLS enabled, returns URL and Close function
+// for the server
+func NewWebhookServer(handler http.Handler) (string, func(), error) {
```
NewAdmissionWebhookServer

sttts commented on test/utils/audit.go:
```diff
@@ -46,6 +48,11 @@ type AuditEvent struct {
+
+	// The Check functions in this package takes ownerships of these maps. You should
+	// not reference these maps after calling the Check functions.
+	WebhookMutationAnnotations map[string]string
```
AdmissionWebhook...

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -97,8 +109,11 @@ func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr admission.Attrib
 		t := time.Now()
-
-		changed, err := a.callAttrMutatingHook(ctx, hook, invocation, versionedAttr, o)
+		round := 1
```
round 0 or 1?

sttts replied:
`mutation` vs. `mutation-2` is maybe best.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -42,12 +43,22 @@ type attributesRecord struct {
 }
 
+// annotation holds a pair of intended audit level and annotation string
+type annotation struct {
```
would remove drop the godoc comments actually. It's just a pair with obvious meaning.

roycaihw committed:
audit & admission: associate annotation with audit level

sttts commented:
Sgtm now. Some nits left.

@liggitt you want to take a last pass?

sttts commented:
/approve

roycaihw committed:
mutating webhook: audit log mutation existence and actual patch

roycaihw committed:
tests

roycaihw committed:
generated

roycaihw commented:
/retest

roycaihw commented:
/retest

roycaihw commented:
/retest

krmayankk commented:
this PR seems like in the write direction, I would love for this to get merged.  Are there any prometheus metrics we are increment per webhook for failed vs successfull ? The easier we make it to debug mutating webhooks , the better :-)

roycaihw commented:
@krmayankk thanks. I was busy with other things. Will try to get back to this probably next week.

We do have metrics counting webhook rejection https://github.com/kubernetes/kubernetes/pull/55183. I'd like to investigate if we have/need metrics to distinguish webhook failure (e.g. server error) v.s. rejection.

roycaihw commented:
Rebased and added reinvocation testcases. PTAL @jpbetz @liggitt @sttts

roycaihw commented:
/retest

roycaihw commented:
gentle ping

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -206,14 +222,28 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta
 		return false, webhookerrors.ToStatusErr(h.Name, result.Result)
 	}
 
+	patchAnnotationKey := PatchAuditAnnotationPrefix + h.Name
```
Name is not globally unique. We need to qualify it at least with the webhook configuration name, e.g.:

https://github.com/kubernetes/kubernetes/blob/661e4d5b3857b5268f2d6d6b79219daf2e4b5833/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/accessors.go#L27

roycaihw replied:
good point. Added webhook config name/UID into consideration

roycaihw commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -41,10 +41,27 @@ import (
+	// Audit handler decides whether annotation with this prefix should be logged based on audit level.
+	// Since mutating webhook patches the request body, audit level must be greater or equal to Request
+	// for the annotation to be logged
+	PatchAuditAnnotationPrefix = "patch.webhook.admission.k8s.io/"
```
I did some research and there are fundamental problems with how the annotation keys should look like. Sorry I couldn't make the following shorter.

The goal is to uniquely capture a webhook invocation event, and the json patch being applied in audit annotations. 

**Background**:

In Kubernetes we have the restriction that an annotation key must be a [qualified name](https://github.com/kubernetes/kubernetes/blob/2974adff27619bb27a27a78bd61de772a2d65bef/staging/src/k8s.io/apiserver/pkg/admission/attributes.go#L191) in format of:
- `{optional DNS subdomain prefix with no more than 253 characters}/{a qualified name that consists of alphanumeric characters, '-', '_' or '.', with no more than 63 characters}`

and the semantics is:
- `{plugin}.{org}/{key}`

**This PR**:

The original idea was to use `webhook.admission.k8s.io` as `{org}` and webhook name as `{key}`, so that we know there will be no key conflict against [custom annotations](https://github.com/kubernetes/kubernetes/blob/2974adff27619bb27a27a78bd61de772a2d65bef/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go#L198-L203), and it's clear from the prefix that the annotation is added by the apiserver (admission webhook org). 

However, [webhook names](https://github.com/kubernetes/kubernetes/blob/2974adff27619bb27a27a78bd61de772a2d65bef/pkg/apis/admissionregistration/validation/validation.go#L261) are allowed to be DNS subdomain with no more than 253 characters, which means webhook name > 63 characters won't get logged. 

**New proposal:**

I propose we use the following semantics for the annotation keys:
- `{webhook name}/k8s_{patch|mutation}_{pass}_{webhook config UID}`

and if in future we want to expose the order of webhook invocation within one pass, we can append the order to the end:
- `{webhook name}/k8s_{patch|mutation}_{pass}_{webhook config UID}_{order}`

Pros: 
- UID is a 36 character string in current [k8s implementation](https://github.com/kubernetes/kubernetes/blob/2974adff27619bb27a27a78bd61de772a2d65bef/staging/src/k8s.io/apimachinery/pkg/util/uuid/uuid.go#L26). We are mostly certain that we won't hit the 63-char limit
- The qualified name segment allows `_`, which is not used by any of the IDs that we are using, makes it easy to parse the name

Cons:
- The prefix is in the webhook name org. Potentially the key name can conflict with a custom annotation (not likely)
- Analyzing the annotations require filtering out the custom annotations, either by doing it manually or through a regex

**Alternatives considered:**
1. We could put less information to the annotation keys and more information to the values, if audit annotations are allowed to be overwritten. Overwriting was designed to be not allowed: https://github.com/kubernetes/kubernetes/pull/58806#discussion_r163921756
2. We can still use `webhook.admission.k8s.io` as `{org}` in the prefix, if we can have an identifier for each webhook which is shorter than the full webhook name (e.g. add a new UID field in the [webhook API](https://github.com/kubernetes/kubernetes/blob/2974adff27619bb27a27a78bd61de772a2d65bef/staging/src/k8s.io/api/admissionregistration/v1/types.go#L302-L309), and have the key be `patch.webhook.admission.k8s.io/{webhook UID}_{webhook config UID}`). Problem with this approach is that it's hard to locate a webhook from the webhook UID, and the UID can be fragile to webhook configuration updates.

@jpbetz @sttts WDYT?

roycaihw replied:
Another idea: use 
- `{patch|mutation}.webhooks.admission.k8s.io/pass_{pass id}_order_{order id}`

as annotation key, and put {`webhook name`, `configuration name` and `existence or patch`} into the annotation value (better if in a structural format that's easy to parse)

The webhook name was put in the key before because we need something unique for each audit event (each invocation), so that we don't violate the no-overwriting rule. I think a combination of `pass id` + `order id` serves the purpose better.

A disadvantage of this approach is that if someone is analyzing a specific webhook's behavior, they cannot just filter the annotation keys, but need to check the annotation values. The extra effort seems reasonable.

jpbetz replied:
Yes, using `webhooks.admission.k8s.io` as the key prefix is what I was expecting.

Having the pass and order field named clearly makes it easy to read the annotations and order them into a chronology of webhook calls so I like how they are presented.

I'm okay with webhook name being in the value.

The only alternative I thought about was doing something like:

key: `*webhooks.admission.k8s.io/_pass_{pass id}`, value: `{webhook_UID_1},{webhook_UID_2}`

But that might be more cumbersome to parse? And I don't know if the value has a size limit.

roycaihw replied:
I didn't see a limit for the value size. I want to structure the value as a JSON payload, so we could aggregate all the webhooks into a list. I don't feel strongly either way.

roycaihw commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -41,10 +41,27 @@ import (
+	PatchAuditAnnotationPrefix = "patch.webhook.admission.k8s.io/"
+	// SecondPassPatchAuditAnnotationPrefix is an audit annotation prefix for webhook patch that happens
+	// in reinvocation
+	SecondPassPatchAuditAnnotationPrefix = "patch-2.webhook.admission.k8s.io/"
```
does https://github.com/kubernetes/kubernetes/pull/77824#discussion_r314835565 look good to you?

jpbetz replied:
Yeah. We can resolve this and continue the discussion there.

roycaihw commented:
@jpbetz Comments about annotation format addressed. Please take a look. Will squash after review

/retest

jpbetz commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher_test.go:
```diff
@@ -0,0 +1,109 @@
+			config:   "test-config",
+			webhook:  "test-webhook",
+			patch:    `[{"op": "add", "path": "/metadata/labels/a", "value": "true"}]`,
+			expected: `{"configuration": "test-config", "webhook": "test-webhook", "patch": "[{\"op\": \"add\", \"path\": \"/metadata/labels/a\", \"value\": \"true\"}]"}`,
```
Would it make sense to inline the patch json instead of escaping it and putting it in a string?

In the future we might support other patch formats. Also include the patch type field to future proof this?

roycaihw replied:
good point. Added patch type field in the payload to future proof other patch formats. Added code to ensure the patch binary is a valid JSON patch and then inline it

jpbetz commented:
Just a couple comments then lgtm.

jpbetz commented:
/lgtm

roycaihw commented:
/hold

adding a couple more JSON patch test cases and squashing

roycaihw commented:
/hold cancel

squashed

jpbetz commented:
/lgtm

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -257,10 +276,76 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta
+	if _, err := jsonpatch.DecodePatch(patch); err != nil {
+		return "", fmt.Errorf("error decoding JSON patch: %v, error: %v", patch, err)
+	}
+	return fmt.Sprintf(`{"configuration": "%s", "webhook": "%s", "patch": %s, "type": "%s"}`, configuration, webhook, patch, admissionv1.PatchTypeJSONPatch), nil
```
we should not add more places where we form json manually. xref https://github.com/kubernetes/kubernetes/issues/81134

liggitt replied:
a small struct that defines the fields and we can pass to json.Marshal() would be safer, e.g.
```go
type MutationAuditAnnotation struct {
  Configuration string      `json:"configuration"`
  Webhook       string      `json:"webhook"`
  Patch         interface{} `json:"patch,omitempty"`
  PatchType     string      `json:"patchType,omitempty"`
}
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -257,10 +276,76 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta
+}
+
+func jsonPatchAnnotationValue(configuration, webhook string, patch []byte) (string, error) {
+	if _, err := jsonpatch.DecodePatch(patch); err != nil {
```
if we get to the point of calling this function, can we not assume the patch is well-formed?

liggitt replied:
also, is there a way to avoid decoding the patch twice?

liggitt replied:
is there a max-allowed length for audit annotations?

roycaihw replied:
We know the patch is well-formed already when we call this function. This is a redundant check. Dropping this check for performance concern. 

> is there a max-allowed length for audit annotations?

I didn't see one for the `value` from the [documentation](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) and the source code:
- [record into admission attribute](https://github.com/kubernetes/kubernetes/blob/d951d19ec4d8ec829e83d146200ea91b664b6fa6/staging/src/k8s.io/apiserver/pkg/admission/attributes.go#L128-L144)
- [record into audit event]( https://github.com/kubernetes/kubernetes/blob/d951d19ec4d8ec829e83d146200ea91b664b6fa6/staging/src/k8s.io/apiserver/pkg/audit/request.go#L218-L231)

liggitt commented:
/hold

sorry for the hold, json construction needs a slight rework so we don't build potentially unsafe json

liggitt commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go:
```diff
@@ -257,10 +276,76 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta
+	changed = !apiequality.Semantic.DeepEqual(attr.VersionedObject, newVersionedObject)
 	trace.Step("Patch applied")
+	if changed {
+		annotator.addPatchAnnotation(result.Patch, result.PatchType)
```
can we pass `patchObj` instead to save decoding the patch, and make `addPatchAnnotation` take a `patch interface{}` parameter, with the requirement that the parameter be json-encodeable?

roycaihw replied:
switched to passing `patchObj` as an `interface{}` and encoding the Go type to JSON

an additional change: log the patch regardless of `changed`

roycaihw commented:
squashed. Thanks for the review

roycaihw commented:
/retest

liggitt commented:
/lgtm
/approve

roycaihw commented:
/hold cancel

roycaihw commented:
/retest

roycaihw commented:
/retest

tedyu commented on staging/src/k8s.io/apiserver/pkg/admission/attributes.go:
```diff
@@ -120,26 +126,36 @@ func (record *attributesRecord) getAnnotations() map[string]string {
 	cp := make(map[string]string, len(record.annotations))
 	for key, value := range record.annotations {
-		cp[key] = value
+		if value.level.Less(maxLevel) || value.level == maxLevel {
```
How about introducing LessOrEqual to combine the checks ?

tedyu commented on staging/src/k8s.io/apiserver/pkg/admission/audit.go:
```diff
@@ -84,11 +84,18 @@ func ensureAnnotationGetter(a Attributes) error {
 	switch a := a.(type) {
 	case privateAnnotationsGetter:
-		audit.LogAnnotations(handler.ae, a.getAnnotations())
+		for key, value := range a.getAnnotations(handler.ae.Level) {
```
These 3 lines can be saved if we move line 95 here

tedyu commented:
@roycaihw 
If you think the above two comments make sense, I can send out a PR.

Thanks

roycaihw commented:
/area admission-control
